File: src/core/SDRMap.py
class SDRMap:

    def __init__(self,size=8,threshold=0.5):
        self.weights = {}
        self.size = size
        self.threshold = threshold

    def set(self,key,val):
        for i in key:
            if(self.weights.get(i) == None):
                self.weights[i] = {}
            for v in val:
                self.weights[i][v] = 1

    def get(self,key):
        sum = {}
        for k in key:
            weights = self.weights.get(k) or {}
            for index, weight in weights.items():
                sum[index] = (sum.get(index) or 0) + weight
        val = []
        for output in sum:
            t = self.size * self.threshold
            if(sum[output] > t):
                val.append(output)
        return val

File: src/core/SDRMap.js

module.exports = class SDRMap {

    constructor(population=8,threshold=0.5){
        this.weights = {}
        this.population = population
        this.threshold = threshold
    }

    set(key,val){
        for(var i = 0; i < key.length; i++){
            this.weights[key[i]] = (this.weights[key[i]] || {})
            for(var j = 0; j < val.length; j++)
                this.weights[key[i]][val[j]] = 1
        }
    }

    get(key){
        const sum = {}
        for(var i = 0; i < key.length; i++){
            const weights = this.weights[key[i]] || {}
            for(var j in weights)
                sum[j] = (sum[j] || 0) + weights[j]
        }
        const val = []
        const threshold = this.population * this.threshold
        for(var i in sum)
            if(sum[i] > threshold)
                val.push(parseInt(i))
        return val.length > 0 ? val : null
    }

    clear(){
        this.weights = {}
    }

}